File: problems/103/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        
        ans = []
        reverse = False
        q = deque([root])

        while q:
            qlen, newQ = len(q), deque()
            ans.append([])

            while qlen:
                qlen -= 1
                node = q.popleft()
                ans[-1].append(node.val)
                
                if node.left:
                    newQ.append(node.left)
                
                if node.right:
                    newQ.append(node.right)
            
            q = newQ
            
            if reverse:
                ans[-1] = ans[-1][::-1]
            
            reverse = not reverse
        
        return ans

File: problems/103/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
    if root == nil {
        return [][]int{}
    }
    
    ans, q, dir := [][]int{}, []*TreeNode{root}, true
    
    for len(q) > 0 {
        curLvl, newQ := []int{}, []*TreeNode{}
        
        for _, node := range q {
            curLvl = append(curLvl, node.Val)
            
            if node.Left != nil {
                newQ = append(newQ, node.Left)
            }
            
            if node.Right != nil {
                newQ = append(newQ, node.Right)
            }
        }
        
        if !dir {
            reverse(curLvl)
        }
        
        ans = append(ans, curLvl)
        q = newQ
        dir = !dir
    }
    
    return ans
}

func reverse(ar []int) {
    lo, hi := 0, len(ar) - 1
    
    for lo < hi {
        ar[lo], ar[hi] = ar[hi], ar[lo]
        lo++
        hi--
    }
}
